### ISSUE TITLE
Support for auto-completion

### ISSUE THREAD

**Issue by samuell** [gopherdata/gophernotes#22]
Is there any way to get auto-completion support (via e.g. [gocode](https://github.com/nsf/gocode))?

If not possible right now, I think it would be an awesome addition.

**Comment by dwhitena** [comment]
Agreed!

**Comment by srhopkins** [comment]
This would be awesome.

**Comment by srhopkins** [comment]
+1

**Comment by graphaelli** [comment]
Spent some time on this -

<img width="366" alt="gophernotes autocomplete" src="https://user-images.githubusercontent.com/83483/31359785-6b46fc10-ad19-11e7-86e2-a8cca3107f8a.png">

https://github.com/gopherdata/gophernotes/compare/master...graphaelli:autocomplete - would welcome a hand and/or some feedback if anyone is interested.

**Comment by dwhitena** [comment]
That's great @graphaelli!  I'm excited to take a look.  It will probably be towards the end of this week though. Feel free to ping me here anything.

**Comment by dwhitena** [comment]
Any more progress on this @graphaelli? Can we help in any particular way?

**Comment by graphaelli** [comment]
Hey @dwhitena.  Sorry about the silence, life etc.  The biggest problem right now is that all of the code for the completion needs to be in a single cell - is there a way to dump the current state of the code accumulated by the notebook?  Once that's set the tests are in place but could use a review There are a bunch of style-type things to implement too, like adding `()` to completions that are functions.

**Comment by dwhitena** [comment]
@graphaelli now worries at all. I totally understand, and no pressure. Just making sure you weren't blocked on me for any reason. 

Regarding dumping the code from the notebook. There was a way to do this before we switched to gomacro, but I don't think there is any more. Is that something that would help? I can open an issue. Also, I will put it on my list of things to get that review done. Thanks again!

**Comment by SpencerPark** [comment]
gomacro has [WriteDeclsToStream](https://github.com/cosmos72/gomacro/blob/05435a41be593c4b39ba680ea7f7f2e7de1852e5/base/global.go#L256) which looks like what we need. 

I haven't tried it and don't know if the format is what you are looking for but hopefully it is a good starting point to look into. The environment to call it on is available in the interpreter (see the code for `handleExecuteRequest`) and assuming things are aptly named I would think this should give you all the code that was previously executed in the notebook.

**Comment by t-0-m-1-3** [comment]
@graphaelli hey if you need an extra set of hands to help out, more than willing to pound keys for you

**Comment by dwhitena** [comment]
That would be great @t-0-m-1-3! Any thought @graphaelli?

**Comment by graphaelli** [comment]
Please have at it, I haven't touched the branch in ages

**Comment by cosmos72** [comment]
While studying how to integrate gophernotes / gomacro with gocode, I found this thread: https://github.com/nsf/gocode/issues/500
where gocode author discusses the problems due to go 1.10 build cache, that strongly reduces the need for "go install", which is instead required for gocode to work.
Despite being almost a de-facto standard, gocode author @nsf discouragingly stated, among other things:

------------
The problem is that a proper tool should work with source files. Hence, there are two options for me:

1. Make gocode work with source files.
2. Rewrite it completely.

Problem: lack of motivation and free time. I definitely favor second option, but it requires way more time than the first one.

I don't know what else to say.

---------

From the technical point of view, I checked what gocode needs: a source file with the cursor position where to perform autocompletion, plus "go install" of all packages to use. And using gomacro [WriteDeclsToStream](https://github.com/cosmos72/gomacro/blob/05435a41be593c4b39ba680ea7f7f2e7de1852e5/base/global.go#L256) seems appropriate.

Any taker?

The alternative would be to use some other code analysis and auto-completion tool.

For example, does anyone have updated information on [guru](http://golang.org/s/using-guru) autocompletion support?  
Visual Studio Code tried to evaluate it 2 years ago, but at the time it seemed too early - see their [issue #288](https://github.com/Microsoft/vscode-go/issues/288)

**Comment by graphaelli** [comment]
That's a bummer, I didn't know about the 1.10 cache affecting this.  Last I heard about this topic on golang-dev was: https://groups.google.com/d/msg/golang-dev/mLsGwoFwqoI/BVSdX7HeBQAJ.  Seems like [sourcegraph](https://github.com/sourcegraph/go-langserver) is worth a shot?  I've never used it.

**Comment by cosmos72** [comment]
I just added a method Interp.CompleteWords() to upstream gomacro.
It returns possible completions in most (not all) cases - if I have time, I will try to merge it into gopherdata in the next days.

Compared to https://github.com/gopherdata/gophernotes/compare/master...jpark-dev:feature/autocomplete, my approach has the advantage of being side-effects-free:
@jpark-dev used `Interp.RunExpr()` for simplicity in his implementation, which unluckily means there are cases where typing things like
```
{ LaunchNuclearMissiles();
```
then pressing TAB may have unwanted side effects - possibly **very** unwanted ones.

**Comment by cosmos72** [comment]
My stab at auto-completion is the branch https://github.com/cosmos72/gophernotes/tree/autocomplete and corresponding pull request https://github.com/gopherdata/gophernotes/pull/111

I am interested in feedback, especially from @jpark-dev, on how it performs compared to https://github.com/gopherdata/gophernotes/compare/master...jpark-dev:feature/autocomplete

**Comment by jpark-dev** [comment]
@cosmos72 It works great! As you pointed out, my implenetation executes the expressions, and I was not a big fan of it. Although I didn't  anyone would launch nuclear missiles, but evaluating long-running expression is not good. (I didn't have a good idea how to do auto-completion without evaluting, and kinda hoped you can add that feature, and it happened! :))
I'd be nice to have completion for function calls(especially since the library I am playing with uses nested calls), but I will happily use work arounds.

BTW, jupyter notebook kinda supports auto completion for local varaibles

![image](https://user-images.githubusercontent.com/4137630/40881925-2ef3a60e-66a1-11e8-975a-57bfa878b49e.png)

(I think the keyword in blue is automatically added by jupyter itself..)

**Comment by cosmos72** [comment]
Good to know, thanks!

Well, the "launch nuclear missile" was an extreme example of side effect :)

I wrote it to point out that, in my opinion, executing an expression when pressing TAB for auto-completion would be unexpected and confusing to programmers.

**Comment by cosmos72** [comment]
added in commit 886f25fc960915616bb97f3867d5ca688125d8e6
